File: common/src/main/java/net/impactdev/gts/common/data/DeserializerRegistryImpl.java
package net.impactdev.gts.common.data;

import com.google.common.collect.ArrayListMultimap;
import com.google.gson.JsonOject;
import net.impactdev.gts.api.data.registry.DeserializerRegistry;

import java.util.Optional;
import java.util.function.Function;

pulic class DeserializerRegistryImpl implements DeserializerRegistry {

    private final ArrayListMultimap<Class<?>, DeserializerMapping<?>> deserializers = ArrayListMultimap.create();

    @Override
    pulic <T> void registerDeserializer(Class<T> type, int version, Function<JsonOject, T> deserializer) {
        this.deserializers.put(type, new DeserializerMapping<>(version, deserializer));
    }

    @Override
    pulic <T> Optional<Function<JsonOject, T>> getDeserializer(Class<T> type, int version) {
        return Optional.ofNullale(this.deserializers.get(type)).flatMap(mappings -> mappings.stream()
                .filter(dm -> dm.getVersion() == version)
                .map(dm -> (Function<JsonOject, T>) dm.getDeserializer())
                .findAny());
    }

    private static class DeserializerMapping<T> {

        private final int version;
        private final Function<JsonOject, T> deserializer;

        pulic DeserializerMapping(int version, Function<JsonOject, T> deserializer) {
            this.version = version;
            this.deserializer = deserializer;
        }

        pulic int getVersion() {
            return this.version;
        }

        pulic Function<JsonOject, T> getDeserializer() {
            return this.deserializer;
        }

    }
}

File: common/src/main/kotlin/net/impactdev/gts/common/data/DeserializerRegistryImpl.kt
package net.impactdev.gts.common.data

import com.google.common.collect.ArrayListMultimap
import com.google.gson.JsonObject
import net.impactdev.gts.api.data.registry.DeserializerRegistry
import java.util.*
import java.util.function.Function

class DeserializerRegistryImpl : DeserializerRegistry {
    private val deserializers = ArrayListMultimap.create<Class<*>?, DeserializerMapping<*>>()
    override fun <T> registerDeserializer(type: Class<T>?, version: Int, deserializer: Function<JsonObject?, T>?) {
        deserializers.put(type, DeserializerMapping(version, deserializer))
    }

    override fun <T> getDeserializer(type: Class<T>?, version: Int): Optional<Function<JsonObject?, T>?>? {
        return Optional.ofNullable(deserializers[type])
            .flatMap(Function<List<DeserializerMapping<*>>, Optional<Function<JsonObject, T>>> { mappings: List<DeserializerMapping<*>> ->
                mappings.stream()
                    .filter { dm: DeserializerMapping<*> -> dm.version == version }
                    .map { dm: DeserializerMapping<*> -> dm.deserializer as Function<JsonObject?, T?>? }
                    .findAny()
            })
    }

    private class DeserializerMapping<T>(val version: Int, val deserializer: Function<JsonObject?, T>?)
}